app/Main.hs

```diff
@@ -35,22 +36,26 @@ import qualified GithubApi
 import qualified Logic
 import qualified Project
 
-getConfigFilePathOrExit :: IO FilePath
-getConfigFilePathOrExit = do
-  args <- System.Environment.getArgs
-  case args of
-    fname : _ -> do
-      exists <- FileSystem.doesFileExist fname
-      if exists then
-        return fname
-      else
-        die $ "Cannot load configuration: the file '" ++ fname ++ "' does not exist."
-    [] ->
-      die $ "No config file specified.\n" ++
-            "The first argument must be the path to the config file."
+data Options = Options
+  { configFilePath :: FilePath
+  , readOnly :: Bool
+  }
+
+commandLineParser :: Opts.ParserInfo Options
+commandLineParser =
+  let
+    optConfigFilePath = Opts.argument Opts.str (Opts.metavar "<config-file>")
+    optReadOnly = Opts.switch (Opts.long "read-only")
+    opts = Options <$> optConfigFilePath <*> optReadOnly
+    help = Opts.fullDesc <> Opts.header "A gatekeeper for your commits"
+  in
+    Opts.info (opts <**> Opts.helper) help
 
 loadConfigOrExit :: FilePath -> IO Configuration
 loadConfigOrExit fname = do
+  exists <- FileSystem.doesFileExist fname
+  unless exists $
+    die $ "Cannot load configuration: the file '" ++ fname ++ "' does not exist."
```

Comment:
What happens if you leave this check out? Just because the file exists and is not a directory doesn’t mean you can read it.

Reply:
It will crash elsewhere, the check is mostly to give a helpful message to the user, not for correctness. It also has a race, because the file could be removed after the check.